src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -171,51 +174,55 @@ internal static bool IsBitwiseEquatable<T>()
-        // points at the number of components, skipping over these two pointer-sized fields.
-        // So substrate those from BaseSize before adding to the RawData.Data reference.
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static unsafe ref byte GetRawArrayData(this Array array) =>
-            ref Unsafe.AddByteOffset(ref Unsafe.As<RawData>(array).Data, (nuint)GetObjectMethodTablePointer(array)->BaseSize - (nuint)(2 * sizeof(IntPtr)));
+            // See comment on RawArrayData for details
+            ref Unsafe.AddByteOffset(ref Unsafe.As<RawData>(array).Data, (nuint)GetMethodTable(array)->BaseSize - (nuint)(2 * sizeof(IntPtr)));
 
         internal static unsafe ushort GetElementSize(this Array array)
         {
             Debug.Assert(ObjectHasComponentSize(array));
-            return GetObjectMethodTablePointer(array)->ComponentSize;
+            return GetMethodTable(array)->ComponentSize;
         }
 
-        // Returns true iff the object has a component size;
-        // i.e., is variable length like System.String or Array.
+        // Returns pointer to the multi-dimensional array bounds.
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal static unsafe bool ObjectHasComponentSize(object obj)
+        internal static unsafe ref int GetMultiDimensionalArrayBounds(Array array)
         {
-            // The Flags field of the method table class will have its high bit set if the
-            // method table has component size info stored somewhere. See member
-            // MethodTable:IsStringOrArray in src\vm\methodtable.h for full details.
-            return (int)GetObjectMethodTablePointer(obj)->Flags < 0;
+            Debug.Assert(GetMultiDimensionalArrayRank(array) > 0);
+            // See comment on RawArrayData for details
+            return ref Unsafe.As<byte, int>(ref Unsafe.As<RawArrayData>(array).Data);
         }
 
-        // Subset of src\vm\methodtable.h
-        [StructLayout(LayoutKind.Explicit)]
-        private struct MethodTable
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe int GetMultiDimensionalArrayRank(Array array)
         {
-            [FieldOffset(0)]
-            public ushort ComponentSize;
-            [FieldOffset(0)]
-            public uint Flags;
-            [FieldOffset(4)]
-            public uint BaseSize;
+            int rank = GetMethodTable(array)->MultiDimensionalArrayRank;
+            GC.KeepAlive(array); // Keep MethodTable alive
+            return rank;
+        }
+
+        // Returns true iff the object has a component size;
+        // i.e., is variable length like System.String or Array.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe bool ObjectHasComponentSize(object obj)
+        {
+            return GetMethodTable(obj)->HasComponentSize;
```

Comment:
Why is `GC.KeepAlive` required in `GetMultiDimensionalArrayRank` but not here?

Reply:
All callers ObjectHasComponentSize keep the MethodTable alive by using the object after the calling ObjectHasComponentSize. I will add a comment or a call to GC.KeepAlive to make this clear.

Reply:
Comment added in https://github.com/dotnet/coreclr/pull/27668